cpp/ql/src/semmle/code/cpp/Element.qll

```diff
@@ -2,13 +2,31 @@ import semmle.code.cpp.Location
 private import semmle.code.cpp.Enclosing
 private import semmle.code.cpp.internal.Type
 
+cached @element resolveElement(@element e) {
+  if isClass(e)
+  then result = resolve(e)
+  else result = e
+}
+
+cached Element mkElement(@element e) {
+  result = resolveElement(e)
+}
+
+cached @element unresolveElement(Element e) {
+  resolveElement(result) = e
+}
+
+cached @element underlyingElement(Element e) {
+  result = e
+}
```

Comment:
Why make all four predicates `cached`? Do they compute more than they seem to because of the casts between `@element` and `Element`? At least `unresolveElement` wouldn't have to be cached if it were defined in terms of `mkElement`, right?

I don't know how the optimizer turns these `cached` annotations into stages, but I'm guessing it will create a stage per predicate here. We can prevent that by grouping the predicates into a `cached private module` and importing that.

Reply:
I did some experiments, and found that performance drop is none are cached, but only `resolveElement` needs to be cached.